### PR TITLE
Allow additional configuration metadata to VMs

### PR DESCRIPTION
**Pull request by anuragaw**
Currently, CloudStack does not allow additional configuration metadata to guest VMs during deployment. This feature aims to provide operators the functionality to pass arbitrary additional configuration metadata that will be included everytime the guest VM is started.

FS document: https://cwiki.apache.org/confluence/display/CLOUDSTACK/Enable+additional+configuration+metadata+to+virtual+machines

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

### CODE REVIEW

**Comment by anuragaw** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by borisstoyanov** [comment]
@blueorangutan test matrix

**Comment by wido** [comment]
This seems to be Xen specific, right?

What can this do in addition to user-data through cloud-init?

**Comment by borisstoyanov** [comment]
This is not just Xen specific but for vmware and kvm as well @wido, it only adds extra configuration to a VM it has nothing to do with user data and cloud-init.

**Comment by wido** [comment]
@borisstoyanov I got confused by some keys I noticed.

I see it also works for KVM, but what kind of keys would somebody want to configure with KVM?

**Comment by anuragaw** [comment]
I'll take a look @borisstoyanov today.

**Comment by svenvogel** [comment]
@anuragaw Hi. Small question. how does this work in practise?

**Comment by anuragaw** [comment]
@blueorangutan package

**Comment by anuragaw** [comment]
@blueorangutan package

**Comment by anuragaw** [comment]
@blueorangutan test

**Comment by anuragaw** [comment]
@blueorangutan package

**Comment by anuragaw** [comment]
@blueorangutan test

**Comment by anuragaw** [comment]
Test failures don't seem related. Please let me know if it doesn't seem right @rhtyd / @borisstoyanov / @shwstppr

**Comment by borisstoyanov** [comment]
@blueorangutan test

File: test/integration/smoke/test_deploy_vm_extra_config_data.py [hunk 1]
```diff
@@ -0,0 +1,529 @@
...
+        '''
+        extraconfig = "%3CmemoryBacking%3E%0D%0A++%3Chugepages%2F%3E%0D%0A%3C%2FmemoryBacking%3E"
+
+        try:
+            # Clear KVM allow list to show that code throws exception when command is not included in the list
+            name = 'allow.additional.vm.configuration.list.kvm'
+
+            self.add_global_config(name, "")
+            self.assertRaises(Exception,
+                              self.deploy_vm(hypervisor, extraconfig),
```
Comment (borisstoyanov) [inline test/integration/smoke/test_deploy_vm_extra_config_data.py]: Screenshot 2019-11-28 at 14 13 53
VMs per each hypervisor are being left behind, perhaps we could add a 'self.destroy_vm(response.id)' same as next tests

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@anuragaw is this ready for review/testing or still work in progress? cc @borisstoyanov

**Comment by anuragaw** [comment]
We have tested this and @andrijapanicsb has found an issue. I'll be fixing it this week.

**Comment by anuragaw** [comment]
Rebased and ready for review @DaanHoogland , @rhtyd , @borisstoyanov , @shwstppr , @nvazquez , @andrijapanicsb 

@blueorangutan package

File: plugins/hypervisors/xenserver/src/main/java/org/apache/cloudstack/hypervisor/xenserver/ExtraConfigurationUtility.java [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+import com.xensource.xenapi.Connection;
+import com.xensource.xenapi.Types;
+import com.xensource.xenapi.VM;
+import org.apache.log4j.Logger;
+import org.apache.xmlrpc.XmlRpcException;
+
+public class ExtraConfigurationUtility {
+    private static final Logger LOG = Logger.getLogger(ExtraConfigurationUtility.class);
+
+    public static void setExtraConfigurationToVm(Connection conn, VM.Record vmr, VM vm, Map<String, String> extraConfig) {
```
Comment (DaanHoogland) [inline plugins/hypervisors/xenserver/src/main/java/org/apache/cloudstack/hypervisor/xenserver/ExtraConfigurationUtility.java]: can you factor out (at least the switches) for readability?

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 3]
```diff
@@ -5105,32 +5130,132 @@ public UserVm createVirtualMachine(DeployVMCmd cmd) throws InsufficientCapacityE
...
+            for (String cfg : extraConfigs) {
+                // Validate cfg against unsupported operations set by admin here
+                String[] allowedKeyList = VmwareAdditionalConfigAllowList.value().split(",");
+                boolean validXenOrVmwareConfiguration = isValidXenOrVmwareConfiguration(cfg, allowedKeyList);
+                String[] paramArray = cfg.split("=");
+                if (validXenOrVmwareConfiguration && paramArray.length == 2) {
+                    try {
+                        userVmDetailsDao.addDetail(vm.getId(), paramArray[0].trim(), paramArray[1].trim(), true);
+                    } catch (ArrayIndexOutOfBoundsException e) {
+                        throw new CloudRuntimeException("Issue occurred during parsing of:" + cfg);
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java]: we are keeping a secret what the issue with the cfg is according to the `ArrayIndexOutOfBoundsException`! better add make sure we log or nest the exception

File: plugins/hypervisors/xenserver/src/main/java/org/apache/cloudstack/hypervisor/xenserver/ExtraConfigurationUtility.java (hunk 2 above)
Comment (anuragaw) [reply plugins/hypervisors/xenserver/src/main/java/org/apache/cloudstack/hypervisor/xenserver/ExtraConfigurationUtility.java]: Done.

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java (hunk 3 above)
Comment (anuragaw) [reply server/src/main/java/com/cloud/vm/UserVmManagerImpl.java]: This was redundant as above if checks length. Cleaned up.

**Comment by anuragaw** [comment]
ping @DaanHoogland for re-review when you have time.

@blueorangutan package

**Comment by anuragaw** [comment]
@blueorangutan test

**Comment by anuragaw** [comment]
Needed rebasing after merge conflicts.

@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by anuragaw** [comment]
@DaanHoogland - can you please take a look at the last commit for code review. Thanks!!

@blueorangutan package

**Comment by anuragaw** [comment]
@blueorangutan test

**Comment by andrijapanicsb** [comment]
Test Name | Steps | Expected Result | Status
-- | -- | -- | --
Add configuration when not enabled | Make sure that enable.additional.vm.configuration   is false and add extra configuration following the KVM example bellow. | VM Details are not being updated   (localcloud)   SBCM5> > update virtualmachine   id=26f88815-5e08-4ab1-ba0b-15658110ab40    extraconfig='%3CmemoryBacking%3E%0A%20%20%3Chugepages%2F%3E%0A%3C%2FmemoryBacking%3E'   filter=details,   {} | Passed
Enable configuration | Set enable.additional.vm.configuration to True and   add extra configuration following the KVM example bellow. | User is able to set configuration   (localcloud) SBCM5> > update virtualmachine   id=26f88815-5e08-4ab1-ba0b-15658110ab40    extraconfig='%3CmemoryBacking%3E%0A%20%20%3Chugepages%2F%3E%0A%3C%2FmemoryBacking%3E'   filter=details,   {     “details": {           "extraconfig-1":   "<memoryBacking>\n    <hugepages/>\n</memoryBacking>"         }, | Passed
KVM extra config | Deploy a VM and set extra config with the   following api call:    deploy virtualmachine … extraconfig='%3CmemoryBacking%3E%0A%20%20%3Chugepages%2F%3E%0A%3C%2FmemoryBacking%3E' | Dump xml the domain on the kvm host and verify   that the extra settings are there        virsh   dumpxml 2   …     <memory   unit='KiB'>524288</memory>     <currentMemory   unit='KiB'>524288</currentMemory>     <memoryBacking>       <hugepages/>     </memoryBacking>   … | Passed
VMware extra config | Deploy a VM   Stop the VM   Update the VM with the following api command and   start it   update virtualmachine id=[vm-id] extraconfig=' hypervisor.cpuid.v0%3DFALSE' | Cat the vmx file on the primary storage and   observe that the specific setting is there      cat "$(esxcli vm process list \| grep i-2-8-VM   \| tail -1 \| awk '{print $3}')" \| grep -w 'hypervisor.cpuid.v0 =   "FALSE"'    hypervisor.cpuid.v0   = "FALSE" | Passed
XenServer extra config | Deploy a VM   Stop the VM   Update the VM with the following api command and   start it   update virtualmachine id=[vm-id]   extraconfig=' HVM-boot-policy%3D%0APV-bootloader%3Dpygrub%0APV-args%3Dhvc0' | On the Xen host us the ‘xe vm-param-get param-name’   command to get the specific parameter and observe it’s been set       xe   vm-param-get param-name=PV-bootloader uuid=f5b6ea53-b9dd-d118-a404-582675cc4345     pygrub       xe   vm-param-get param-name=PV-args uuid=f5b6ea53-b9dd-d118-a404-582675cc4345    hvc0 | Passed
KVM key list restriction | Remove all items from allow.additional.vm.configuration.list.kvm   Deploy a VM and set extra config with the   following api call:    deploy virtualmachine …   extraconfig='%3CmemoryBacking%3E%0A%20%20%3Chugepages%2F%3E%0A%3C%2FmemoryBacking%3E' | Exception appears that the item is not allowed.   Any of the child records of that xml node were not added as well       CloudstackAPIException:   Execute cmd: deployvirtualmachine failed, due to: errorCode: 530,   errorText:Extra config  is not on the   list of allowed keys for KVM hypervisor hosts | Passed
VMware key list restriction | Remove all items from allow.additional.vm.configuration.list.vmware   global setting   Deploy a VM   Stop the VM   Update the VM with the following api command and   start it   update virtualmachine id=[vm-id] extraconfig='hypervisor.cpuid.v0%3DFALSE' | Exception appears that the item is not allowed.      CloudstackAPIException: Execute cmd:   deployvirtualmachine failed, due to: errorCode: 530, errorText:Extra config   hypervisor.cpuid.v0=FALSE is not on the list of allowed keys for VMware   hypervisor hosts. | Passed
XenServer key list restriction | Remove all items from allow.additional.vm.configuration.list.xenserver   global setting   Deploy a VM   Stop the VM   Update the VM with the following api command and   start it   update virtualmachine id=[vm-id] extraconfig='HVM-boot-policy%3D%0APV-bootloader%3Dpygrub%0APV-args%3Dhvc0' | Exception appears that the item is not allowed.      CloudstackAPIException: Execute cmd:   deployvirtualmachine failed, due to: errorCode: 530, errorText:Extra config   HVM-boot-policy= is not on the list of allowed keys for XenServer hypervisor   hosts. | Passed
XenServer key mapping negative | Pass an argument that is not valid as per the   given pattern:       Test value:    HVM-boot-policy=   PV-bootloader=pygrub=failure   PV-args=hvc0 | Cloudstack throws an exception when validating the   map:       (localcloud)   SBCM5> > update virtualmachine id=afeb87cf-5a0a-4807-9c69-8644544ccdf9   extraconfig=HVM-boot-policy%3D%0APV-bootloader%3Dpygrub%3Dfailure%0APV-args%3Dhvc0   🙈 Error:   (HTTP 530, error code 4250) The passed extra config string 'HVM-boot-policy=   PV-bootloader=pygrub=failure   PV-args=hvc0'   contains an invalid key/value pair pattern | Passed
Clean extra config data | Stop the VM    Execute update VM with ‘cleanupdetails=true’   argument | Any additional config has been removed       (localcloud)   SBCM5> > list virtualmachines id=9207b8b4-6d94-4a93-ba44-20ad5cd39f76   filter=details,   {     "virtualmachine": [         "details": {           "deployvm":   "true",           "extraconfig-1":   "<memoryBacking>\n    <hugepages/>\n</memoryBacking>"   }   (localcloud)   SBCM5> > update virtualmachine id=9207b8b4-6d94-4a93-ba44-20ad5cd39f76   cleanupdetails=true   {     "virtualmachine": {...}   (localcloud)   SBCM5> > list virtualmachines id=9207b8b4-6d94-4a93-ba44-20ad5cd39f76   filter=details,   {     "virtualmachine": [             "details": {}   } | Passed
KVM – Nested keys not allowed | Create an xml that is containing a not allowed key   inside an allowed key and try submitting it. | (localcloud)   SBCM5> > update virtualmachine   id=9d5b6561-ee7b-4711-94c7-c79f268cfa24    extraconfig='%3CmemoryBacking%3E%0A%20%20%3Chugepages%2F%3E%0A%3C%2FmemoryBacking%3E'   🙈 Error:   (HTTP 530, error code 4250) Extra config hugepages is not on the list   of allowed keys for KVM hypervisor hosts   (localcloud)   SBCM5> > list configurations   name=allow.additional.vm.configuration.list.kvm   {         "name":   "allow.additional.vm.configuration.list.kvm",         "value": "memoryBacking"   } | Passed
Apply configuration when it’s disabled | Set enable.additional.vm.configuration to false   Apply any configurations to a VM. | (localcloud)   SBCM5> > update virtualmachine   id=9d5b6561-ee7b-4711-94c7-c79f268cfa24    extraconfig='%3CmemoryBacking%3E%0A%3C%2FmemoryBacking%3E'   🙈 Error: (HTTP 431, error code 4350) attempted   setting extraconfig but enable.additional.vm.configuration is disabled | Passed
Update some of the enabled configurations | Have a list of enabled keys, but pass just one of   them | (localcloud)   SBCM5> > update virtualmachine   id=9d5b6561-ee7b-4711-94c7-c79f268cfa24    extraconfig='%3CmemoryBacking%3E%0A%3C%2FmemoryBacking%3E'   {     "virtualmachine": {   …       "details": {           "Message.ReservedCapacityFreed.Flag": "true",         "extraconfig-1":   "<memoryBacking>\n</memoryBacking>"       }   …   (localcloud)   SBCM5> > list configurations name=allow.additional.vm.configuration.list.kvm   {         "name":   "allow.additional.vm.configuration.list.kvm",         "value": "memoryBacking,   hugepages"   } | Passed

**Comment by pavanaravapalli** [comment]
deployVirtualMachine API is tested with extraConfig? as i have seen extraconfig value is not persisted in vm details table for deploy vm, yielding no change in the vm config.

**Comment by harikrishna-patnala** [comment]
@pavanaravapalli please raise an issue when you see anything not working with appropriate information and steps. Thanks

**Comment by pavanaravapalli** [comment]
@harikrishna-patnala  ok i will do necessary if required.

**Pull request merged**